pkg/reconciler/subscription/subscription_test.go

```diff
@@ -0,0 +1,700 @@
+	channelURI = "http://" + channelDNS + "/"
+
+	subscriberDNS = "subscriber.mynamespace.svc." + utils.GetClusterDomainName()
+	subscriberURI = "http://" + subscriberDNS + "/"
+
+	replyDNS = "reply.mynamespace.svc." + utils.GetClusterDomainName()
+	replyURI = "http://" + replyDNS + "/"
+
+	serviceDNS = serviceName + "." + testNS + ".svc." + utils.GetClusterDomainName()
+	serviceURI = "http://" + serviceDNS + "/"
+
+	subscriberGVK = metav1.GroupVersionKind{
+		Group:   "testing.eventing.knative.dev",
+		Version: "v1alpha1",
+		Kind:    "Subscriber",
+	}
+
+	serviceGVK = metav1.GroupVersionKind{
+		Version: "v1",
+		Kind:    "Service",
+	}
+
+	channelGVK = metav1.GroupVersionKind{
+		Group:   "eventing.knative.dev",
+		Version: "v1alpha1",
+		Kind:    "Channel",
+	}
+)
+
+func init() {
+	// Add types to scheme
+	_ = eventingv1alpha1.AddToScheme(scheme.Scheme)
+	_ = duckv1alpha1.AddToScheme(scheme.Scheme)
+}
+
+func TestAllCases(t *testing.T) {
+	table := TableTest{
+		{
+			Name: "bad workqueue key",
+			// Make sure Reconcile handles bad keys.
+			Key: "too/many/parts",
+		}, {
+			Name: "key not found",
+			// Make sure Reconcile handles good keys that don't exist.
+			Key: "foo/not-found",
+			//}, { // TODO: there is a bug in the controller, it will query for ""
+			//	Name: "incomplete subscription",
+			//	Objects: []runtime.Object{
+			//		NewSubscription(subscriptionName, testNS),
+			//	},
+			//	Key:     "foo/incomplete",
+			//	WantErr: true,
+			//	WantEvents: []string{
+			//		Eventf(corev1.EventTypeWarning, "ChannelReferenceFetchFailed", "Failed to validate spec.channel exists: s \"\" not found"),
+			//	},
+		}, {
+			Name: "subscription, but subscriber is not addressable",
+			Objects: []runtime.Object{
+				testing2.NewSubscription(subscriptionName, testNS,
+					testing2.WithSubscriptionChannel(channelGVK, channelName),
+					testing2.WithSubscriptionSubscriberRef(subscriberGVK, subscriberName),
+				),
+				testing2.NewUnstructured(subscriberGVK, subscriberName, testNS),
+				testing2.NewChannel(channelName, testNS,
+					testing2.WithInitChannelConditions,
+					testing2.WithChannelAddress(channelDNS),
+				),
+			},
+			Key:     testNS + "/" + subscriptionName,
+			WantErr: true,
+			WantEvents: []string{
+				Eventf(corev1.EventTypeWarning, "SubscriberResolveFailed", "Failed to resolve spec.subscriber: status does not contain address"),
+			},
+			WantStatusUpdates: []clientgotesting.UpdateActionImpl{{
+				Object: testing2.NewSubscription(subscriptionName, testNS,
+					testing2.WithSubscriptionChannel(channelGVK, channelName),
+					testing2.WithSubscriptionSubscriberRef(subscriberGVK, subscriberName),
+					// The first reconciliation will initialize the status conditions.
+					testing2.WithInitSubscriptionConditions,
+				),
+			}},
+		}, {
+			Name: "subscription, but subscriber does not exist",
+			Objects: []runtime.Object{
+				testing2.NewSubscription(subscriptionName, testNS,
```

Comment:
Can you explain why you replaced the previous builder syntax with the functional options syntax?

Reply:
serving uses mutation functions and go tends to not use builders. The functions are more flexible.